aws-cpp-sdk-core/source/utils/threading/Executor.cpp

```diff
@@ -107,3 +107,58 @@ bool PooledThreadExecutor::HasTasks()
     std::lock_guard<std::mutex> locker(m_queueLock);
     return m_tasks.size() > 0;
 }
+
+BlockingExecutor::BlockingExecutor(std::shared_ptr<Executor> executor, size_t poolSize) :
+m_executor(executor), m_poolSize(poolSize), m_numTasksRunning(0)
+{
+}
+
+BlockingExecutor::~BlockingExecutor()
+{
+    while (m_numTasksRunning != 0)
+    {
+        std::this_thread::yield();
+        if (m_numTasksRunning == 0)
+        {
+            // Make sure the condition actually does hold under the lock
+            std::lock_guard<std::recursive_mutex> locker(m_syncPointLock);
+            if (m_numTasksRunning == 0)
+            {
+                break;
+            }
+        }
+    }
+    assert(m_numTasksRunning == 0);
+}
+
+bool BlockingExecutor::SubmitToThread(std::function<void()>&& fn)
+{
+    std::unique_lock<std::recursive_mutex> locker(m_syncPointLock);
+    assert(m_numTasksRunning <= m_poolSize);
+    if (m_numTasksRunning >= m_poolSize)
+    {
+        m_syncPoint.wait(locker, [&]{
+            return m_numTasksRunning < m_poolSize;
+        });
+    }
+    m_numTasksRunning++;
+    locker.unlock();
+    
+    m_executor->Submit(&BlockingExecutor::ExecuteTask, this, fn);
+    return true;
+}
+
+void BlockingExecutor::ExecuteTask(std::function<void()> fn)
+{
+    fn();
+    OnTaskComplete();
+}
+
+void BlockingExecutor::OnTaskComplete()
+{
+    std::unique_lock<std::recursive_mutex> locker(m_syncPointLock);
+    assert(m_numTasksRunning <= m_poolSize);

```

Comment:
we were hitting this before when the queueing was incorrect, so having an assert seemed prudent.